util/histogram/moving_average_histogram.go

```diff
@@ -0,0 +1,152 @@
+// minute, as well as weighted statistics over the duration, with more recent time periods weighted
+// more heavily.
+type MovingAverageHistogram struct {
+	// typ is the type of duration histogram for this moving average histogram. If necessary, this
+	// can be extended to a description of the bins directly.
+	typ DurationHistogramType
+	// maxDuration is the total history that is kept in the moving average histogram.
+	maxDuration time.Duration
+	// durationPerHistogram describes the rotation schedule of the histogram.
+	durationPerHistogram time.Duration
+
+	// mu is _only_ used to protect rotation of the histograms. Once a reference to a histogram is
+	// held, mu can be released.
+	mu sync.Mutex
+	// h keeps the histograms. h[0] is the active histogram, and histograms get older as they are
+	// farther in the slice.
+	h []*DurationHistogram
+
+	// doneCh is closed to stop automatically rotating
+	doneCh chan struct{}
+}
+
+func NewMovingAverageHistogram(
+	typ DurationHistogramType,
+	maxDuration time.Duration,
+	durationPerHistogram time.Duration,
+) (*MovingAverageHistogram, error) {
+	histCount := maxDuration / durationPerHistogram
+	if maxDuration%durationPerHistogram != 0 {
+		return nil, errors.E("NewMovingAverageHistogram", errors.K.Invalid, "reason", "maxDuration must be a multiple of durationPerHistogram")
+	}
+
+	mah := &MovingAverageHistogram{
+		typ:                  typ,
+		maxDuration:          maxDuration,
+		durationPerHistogram: durationPerHistogram,
+		h:                    make([]*DurationHistogram, histCount),
+	}
+	// Create the first histogram
+	mah.h[0] = NewDurationHistogram(typ)
+
+	return mah, nil
+}
+
+func (m *MovingAverageHistogram) Rotate() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	for i := len(m.h) - 1; i > 0; i-- {
+		m.h[i] = m.h[i-1]
+	}
+
+	m.h[0] = NewDurationHistogram(m.typ)
+}
+
+func (m *MovingAverageHistogram) autoRotate(d time.Duration) {
+	t := time.NewTicker(d)
+
+	for {
+		select {
+		case <-t.C:
+			m.Rotate()
+		case <-m.doneCh:
+			return
+		}
+	}
+}
+
+func (m *MovingAverageHistogram) Start() {
+	go m.autoRotate(m.durationPerHistogram)
+}
+
+func (m *MovingAverageHistogram) Stop() {
+	close(m.doneCh)
+}
+
+func (m *MovingAverageHistogram) Observe(n time.Duration) {
+	m.mu.Lock()
+	h := m.h[0]
+	m.mu.Unlock()
+
+	h.Observe(n)
+}
+
+func (m *MovingAverageHistogram) StatLastMinute(f func(h *DurationHistogram) time.Duration) time.Duration {
+	// We calculate this countToKeep in order to at least capture the last minute worth of data.
+	countToKeep := int(time.Minute / m.durationPerHistogram)
+	if time.Minute%m.durationPerHistogram != 0 {
+		countToKeep++
+	}
+	if countToKeep == 1 {
+		countToKeep++
+	}
+
+	m.mu.Lock()
+	hists := make([]*DurationHistogram, countToKeep)
+	copy(hists, m.h)
+	m.mu.Unlock()
+
+	agg := NewDurationHistogram(m.typ)
+	for _, h := range hists {
+		// Ignore the error as we know that the histograms are of the same type.
+		agg.Add(h)
```

Comment:
```suggestion
		_ = agg.Add(h)
```

also what if h is nil ?